java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java

```diff
@@ -0,0 +1,138 @@
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    if (tree.is(Tree.Kind.CLASS)) {
+      checkFields((ClassTree) tree);
+    } else if (tree.is(Tree.Kind.METHOD)) {
+      checkMethod((MethodTree) tree);
+    }
+  }
+
+  private void checkFields(ClassTree tree) {
+    JavaSymbol.TypeJavaSymbol typeJavaSymbol = (JavaSymbol.TypeJavaSymbol) tree.symbol();
+    typeJavaSymbol.memberSymbols().stream().filter(SymbolUtils::isField).map(Symbol::declaration).map(VariableTree.class::cast)
+      .forEach(declaration -> getIssueMessage(declaration).ifPresent(message -> reportIssue(declaration.type(), message)));
+  }
+
+  private void checkMethod(MethodTree method) {
+    for (VariableTree parameter : method.parameters()) {
+      getIssueMessage(parameter).ifPresent(s -> reportIssue(parameter.type(), s));
+    }
+  }
+
+  private Optional<String> getIssueMessage(VariableTree variable) {
+    if (!validationEnabled(variable) && validationSupported(variable)) {
+      return Optional.of(MessageFormat.format("Add missing \"@Valid\" on \"{0}\" to validate it with \"Bean Validation\".", variable.simpleName()));
+    }
+    return Optional.empty();
+  }
+
+  private boolean validationEnabled(VariableTree variable) {
+    Type type = variable.symbol().type();
+    boolean result = variable.symbol().metadata().isAnnotatedWith(JAVAX_VALIDATION_VALID);
+    if (type instanceof ParametrizedTypeJavaType) {
+      result = result || typeArgumentAnnotationNames(variable).anyMatch(name -> name.equals(JAVAX_VALIDATION_VALID));
+    }
+    return result;
+  }
+
+  private Stream<String> typeArgumentAnnotationNames(VariableTree variable) {
+    return typeArgumentTypeTrees(variable).flatMap(type -> type.annotations().stream()).map(annotation -> annotation.symbolType().fullyQualifiedName());
+  }
+
+  private Stream<TypeTree> typeArgumentTypeTrees(VariableTree variable) {
+    ParameterizedTypeTree parameterizedType = (ParameterizedTypeTree) variable.type();
+    return parameterizedType.typeArguments().stream().map(TypeTree.class::cast);
+  }
+
+  private boolean validationSupported(VariableTree variable) {
+    return annotationInstances(variable).anyMatch(this::isConstraintAnnotation);
+  }
+
+  private Stream<SymbolMetadata.AnnotationInstance> annotationInstances(VariableTree variable) {
+    if (variable.type().is(Tree.Kind.PARAMETERIZED_TYPE)) {
+      return typeArgumentAnnotationInstances(variable);
+    } else {
+      JavaSymbol.TypeJavaSymbol classSymbol = ((JavaType) variable.symbol().type()).getSymbol();
+      return classAndFieldAnnotationInstances(classSymbol);
+    }
+  }
+
+  private Stream<SymbolMetadata.AnnotationInstance> typeArgumentAnnotationInstances(VariableTree variable) {
+    return typeArgumentTypeTrees(variable).map(this::typeSymbol).flatMap(this::classAndFieldAnnotationInstances);
+  }
+
+  private JavaSymbol.TypeJavaSymbol typeSymbol(TypeTree type) {
```

Comment:
I don't understand, isn't that precisely what I already do? I call `symbolType()` on `type` (which is a `TypeTree`) to get a `JavaSymbol.TypeJavaSymbol`.

Reply:
Ah, sorry if you misunderstood. My point was that `JavaSymbol.TypeJavaSymbol` is an implementation class, and I don't see any need to rely on implementation class rather than the interfaces `Symbol` and `Symbol.TypeSymbol`.

I reworked your code on my machine to check that there was no dependency over some implementation-hidden methods, and it's not the case.

So you can rewrite the method above like this, and drop this `typeSymbol(TypeTree)` method

```
private static Stream<SymbolMetadata.AnnotationInstance> typeArgumentAnnotationInstances(VariableTree variable) {
    return typeArgumentTypeTrees(variable)
      .map(TypeTree::symbolType)
      .map(Type::symbol)
      .flatMap(MissingBeanValidationCheck::classAndFieldAnnotationInstances);
  }
```